include/equation_system.hpp

```diff
@@ -0,0 +1,100 @@
+#ifndef ADJACENT_EQUATION_SYSTEM_HPP
+#define ADJACENT_EQUATION_SYSTEM_HPP
+
+#include <vector>
+#include <unordered_map>
+
+#include <xtensor/xtensor.hpp>
+#include "expression.hpp"
+#include "expression_vector.hpp"
+#include "gaussian_method.hpp"
+
+enum SolveResult {
+	OKAY,
+	DIDNT_CONVERGE,
+	REDUNDANT,
+	POSTPONE
+};
+
+using expr_ptr = std::shared_ptr<Expr>;
+
+class EquationSystem  {
+
+public:
+	bool is_dirty = true;
+	int max_steps = 20;
+	int drag_steps = 3;
+	bool revert_when_not_converged = true;
+
+	std::string stats;
+	bool dof_changed;
+
+
+	xt::xtensor<std::shared_ptr<Expr>, 2> J;
```

Comment:
may be use Eigen stuctures since we very need to link it to have awesome perf and as many linear solvers as we want.